src/Patches/ItemRandomizer.cs

```diff
@@ -110,95 +110,135 @@ public static void RandomizeAndPlaceItems(Random random = null) {
-            if (SaveFile.GetInt(SaveFlags.LaurelsLocation) != 0) {
-                foreach (Reward item in ProgressionRewards) {
-                    if (item.Name == "Hyperdash") {
-                        foreach (Location location in InitialLocations) {
-                            if ((location.LocationId == "Well Reward (6 Coins)" && SaveFile.GetInt(SaveFlags.LaurelsLocation) == 1)
-                                || (location.LocationId == "Well Reward (10 Coins)" && SaveFile.GetInt(SaveFlags.LaurelsLocation) == 2)
-                                || (location.LocationId == "waterfall" && SaveFile.GetInt(SaveFlags.LaurelsLocation) == 3)) {
-                                Check Check = new Check(item, location);
-                                string DictionaryId = Check.CheckId;
-                                ProgressionLocations.Add(DictionaryId, Check);
-                                InitialLocations.Remove(location);
-                                ProgressionRewards.Remove(item);
-                                break;
-                            }
+            foreach (Reward item in ProgressionRewards.ToList()) {
+                if (item.Name == "Hyperdash" && SaveFile.GetInt(LaurelsLocation) != 0) {
+                    foreach (Location location in InitialLocations.ToList()) {
+                        if ((location.LocationId == "Well Reward (6 Coins)" && SaveFile.GetInt(LaurelsLocation) == 1)
+                            || (location.LocationId == "Well Reward (10 Coins)" && SaveFile.GetInt(LaurelsLocation) == 2)
+                            || (location.LocationId == "waterfall" && SaveFile.GetInt(LaurelsLocation) == 3)) {
+                            Check Check = new Check(item, location);
+                            string DictionaryId = Check.CheckId;
+                            ProgressionLocations.Add(DictionaryId, Check);
+                            InitialLocations.Remove(location);
+                            ProgressionRewards.Remove(item);
+                            break;
+                        }
+                    }
+                }
+            }
+
+            // pre-place hexagons if keys behind bosses is on
+            foreach (Reward item in InitialRewards.ToList()) {
+                if (item.Name.Contains("Hexagon") && SaveFile.GetInt(KeysBehindBosses) == 1) {
+                    foreach (Location location in InitialLocations.ToList()) {
+                        Check Check = new Check(item, location);
+                        if (Check.CheckId == "Vault Key (Red) [Fortress Arena]"
+                            || Check.CheckId == "Hexagon Green [Library Arena]"
+                            || Check.CheckId == "Hexagon Blue [ziggurat2020_3]") {
+                            ProgressionLocations.Add(Check.CheckId, Check);
+                            InitialLocations.Remove(location);
+                            InitialRewards.Remove(item);
+                            break;
```

Comment:
Could probably combine these into one foreach

Reply:
They're iterating through different lists

Reply:
Wait isn't this turning some of the non-progression-marked gold hexes into progression-marked ones then? Not that it _really_ matters that much, just a tiny bit inefficient

Reply:
Don't they all get marked as progression no matter what? These ones are just replacing the rgb hexes which are normally not progresion, so I think it's the opposite? but they have to get pre-placed anyway so I don't think it matters that much (might be wrong though)

Reply:
So, let's say you have ability shuffle on and 20 hexes to goal, 30 in the pool. It'll mark 15 of the hexes as progression (since you need 15 of them to get all of your abilities), and the remaining 15 will be placed randomly.

With the way the PR is written currently, it takes the gold hexes out of the non-progression pool and places them on bosses, then "marks" them as progression (puts them in the progression locations list). So it'll actually place only 12 of them randomly.

Actually, talking this through, it doesn't matter at all lol. It's a difference between you being 100% expected to grab the hexes on the bosses to unlock all abilities vs having 3 "extras"